rclcpp/include/rclcpp/publisher_options.hpp

```diff
@@ -0,0 +1,158 @@
+};
+
+/// Structure containing configuration options for Publishers
+template<typename Alloc = std::allocator<void>>
+class PublisherOptions
+{
+public:
+  /// Create PublisherOptions with default values, optionally specifying the allocator to use.
+  /**
+   * Default values for the node options:
+   *
+   *   - context = rclcpp::contexts::default_context::get_global_default_context()
+   *   - arguments = {}
+   *   - initial_parameters = {}
+   *   - use_global_arguments = true
+   *   - use_intra_process_comms = false
+   *   - start_parameter_services = true
+   *   - start_publisher_publisher = true
+   *   - publisher_qos_profile = rmw_qos_profile_default
+   *   - allocator = rcl_get_default_allocator()
+   *
+   * \param[in] allocator allocator to use in construction of PublisherOptions.
+   */
+  RCLCPP_PUBLIC
+  PublisherOptions() = default;
+
+  RCLCPP_PUBLIC
+  explicit PublisherOptions(const rmw_qos_profile_t & qos_profile)
+  : qos_profile_(qos_profile) {}
+
+  /// Destructor.
+  RCLCPP_PUBLIC
+  virtual
+  ~PublisherOptions() = default;
+
+  /// Copy constructor.
+  RCLCPP_PUBLIC
+  PublisherOptions(const PublisherOptions & other) = default;
+
+  /// Assignment operator.
+  RCLCPP_PUBLIC
+  PublisherOptions &
+  operator=(const PublisherOptions & other) = default;
+
+  /// Return a reference to the publisher_qos_profile QoS.
+  RCLCPP_PUBLIC
+  const rmw_qos_profile_t &
+  qos_profile() const
+  {
+    return qos_profile_;
+  }
+
+  RCLCPP_PUBLIC
+  rmw_qos_profile_t &
+  qos_profile()
+  {
+    return qos_profile_;
+  }
+
+  /// Set the publisher_qos_profile QoS, return this for chaining
+  /**
+   * The QoS settings to be used for the publisher
+   */
+  RCLCPP_PUBLIC
+  PublisherOptions &
+  qos_profile(const rmw_qos_profile_t & publisher_qos_profile)
+  {
+    qos_profile_ = publisher_qos_profile;
+    return *this;
+  }
+
+  RCLCPP_PUBLIC
+  size_t &
+  qos_history_depth()
+  {
+    return qos_profile_.depth;
+  }
+
+  RCLCPP_PUBLIC
+  PublisherOptions &
+  qos_history_depth(size_t depth)
+  {
+    qos_profile_.depth = depth;
+    return *this;
+  }
+
+
+  RCLCPP_PUBLIC
+  const PublisherEventCallbacks &
+  event_callbacks() const
+  {
+    return callbacks_;
+  }
+
+  /// Return the rcl_allocator_t to be used.
+  RCLCPP_PUBLIC
+  std::shared_ptr<Alloc>
+  allocator() const
+  {
+    return allocator_;
+  }
+
+  /// Set the rcl_allocator_t to be used, may cause deallocation of existing rcl_publisher_options_t
+  /**
+   * This will cause the internal rcl_publisher_options_t struct to be invalidated.
+   */
+  RCLCPP_PUBLIC
+  PublisherOptions &
+  allocator(std::shared_ptr<Alloc> allocator)
+  {
+    allocator_ = allocator;
+    return *this;
+  }
+
+private:
+  // IMPORTANT: if any of these default values are changed, please update the
+  // documentation in this class.
+
+  PublisherEventCallbacks callbacks_;
+
+  rmw_qos_profile_t qos_profile_ = rmw_qos_profile_default;
```

Comment:
Is `rmw_qos_profile_default` a value? If yes, it should be constant and should be written `RMW_QOS_PROFILE_DEFAULT`.

Generally to avoid static initialization nightmare, I'd avoid the global and call a named constructor there.
`rmw_qos_profile_t qos_profile = QosProfile::build_default_profile();`

... or something like that.